### ISSUE TITLE
Reconnecting after connection_lost

### ISSUE THREAD

**Issue by perry-blueberry** [eclipse/paho.mqtt.cpp#241]
I have what seems like a simple question, but I haven't been able to figure out how to do this.

If, for example, the MQTT broker goes down and then up again, I want to be able to reconnect to it again without early termination due to _mqtt::exception_. I know that the callback for this is ```connection_lost``` and my callback is called after the broker goes down. But even when setting ```set_automatic_reconnect``` to true the application crashes once the callback has been run. I have tried to do a reconnect directly in the connection_lost callback, but that just hangs the application. I also suspected it to be because the client is in some intermediate state between disconnected and connected since the broker doesn't say that the client has disconnected, so I tried a ```disconnect``` which seems to crash the application.

Any help would be appreciated.

**Comment by fpagliughi** [comment]
I haven't used automatic reconnect in any of my projects. That was implemented in the underlying Paho C lib, and I just passed through the options. If it isn't working as expected, we should definitely figure it out and create an issue on that repo.

But I do use `connection_lost` all the time. For that, though, you do need to use the Async client and _be sure_ not to make a blocking call inside the function, particularly not a blocking call to the MQTT library. So don't call a `Token::wait()` or anything like that.

The callbacks are all called from the processing thread in the C lib. Blocking it to wait for the library to do something is a deadlock.

**Comment by FanOfSilence** [comment]
When you say you use ```connection_lost``` do you mean in the sense to connect to the broker again once a connection is possible? Because I don't see how that would be possible to do since, as you say, waiting for the library seems to result in a deadlock for some functions once ```connection_lost``` has been reached.

Another question I have is for a possible workaround how to connect again. I was thinking that instead doing a ```reconnect``` a fresh ```connect``` again would work fine if the connection went down. The problem I'm having implementing that is that ```connect``` throws an exception that I am unable to catch even when doing ```catch(...)``` on the call. And I've also tried printing directly in Paho cpp where it tries to connect to see if an exception was thrown there, but it doesn't seem to throw the exception at that point. So my best guess is that the exception is thrown in some callback that is not possible to handle since ```connect``` is the function setting up the callbacks (a bit of a Catch 22 if that is the case). Is that understanding correct or is it thrown somewhere else? Is there some way to catch this exception?

**Comment by fpagliughi** [comment]
There are two clients in the library: the synchronous `client` and the asynchronous `async_client`. You haven't mentioned which you are using.

Methods in the synchronous client block waiting for a response. In the case of connect, QoS 1 and 2 messages, etc, it waits for the ack message back from the broker before returning. It is inappropriate to use these in a callback, because the callback thread is the one that processes incoming messages, like the ack you're waiting for. So deadlock occurs.

The asynchronous client just queues outbound messages and returns a `token` back to the caller which can be used to track progress of an operation. It is safe to use these methods in a callback, but it is not safe to wait for the token to complete by calling _wait()_ on it. In that case, you can set up success/failure callbacks to fire when the operation has completed or failed. On a failure, you just try again, possibly backing off in time a little. The `asunc_subscribe.cpp` sample demonstrates some of this.

You can do asynchronous things directly in a callack, or you can use the callback to signal another thread to take action using the typical C++ thread synchronization mechanisms. Once you pass the action to another thread, you can make blocking calls at your leisure over there.

I should note that there's one interesting exception. In an `on_failure` callback from a connect attempt, you can usually block the callback thread in a sleep() before attempting a reconnect. Since you know that you're not connected (the connect attempt just failed), the callback thread has nothing to do, so you can wait a little while before sending an async reconnect request.

The synchronous client just wraps an async one, so you can use callbacks with one, but that often just gets you into trouble with deadlocks. The more common usage is to detect a disconnect on a failed publish or consume operation and try to reconnect then. The `sync_consume.cpp` example shows this.

You don't need to use `reconnect()`. It's just a convenience function. It just uses a cached copy of your connect options to call `connect()` again. Either way, they just put a CONNECT message out on the wire.

I do assume that you're using the synchronous client if you're getting exceptions from connect(). The `async_client` doesn't typically throw exceptions when initiating an action.. . only when waiting for the token to complete. If you're trying to do some precise actions, you'll get better control using the asynchronous client. Have a look at it.

**Comment by FanOfSilence** [comment]
I am using ```async_client```.
I have looked in ```async_subscribe.cpp``` and it is very similar to what I've tried before.

For example, using ```reconnect()``` from the library, I get a deadlock where nothing happens even if I bring up the MQTT broker again.
I've also tried catching the exception when calling ```connect``` in the library, as I mentioned above. I copied the code from ```async_subscribe.cpp``` function ```reconnect``` but no exception is caught there; it only crashes with ```mqtt::exception```. 
I'm also using ```on_failure``` and that callback is not called either.

**Comment by fpagliughi** [comment]
Have you tried running `async_subscribe` with your setup? Does it exhibit the same behavior? Either way can you post a minimal example that exhibits the problem. I can try to recreate it.

**Comment by FanOfSilence** [comment]
I managed to get a version working based on ```async_subscribe``` (by manually connecting every time the connection is lost from ```on_failure```, since using ```reconnect``` causes a deadlock and ```set_automatic_reconnect(true)``` throws exceptions that I am unable to catch).
However, when I am publishing to MQTT while the connection is down, my application still crashes with ```mqtt::exception MQTT error [-3]: Disconnected```. Shouldn't this be caught in ```on_failure``` like other errors?
I'm having trouble coming up with a minimal example but my ```on_failure``` looks like this:
```
void tMqttClient::on_failure(const mqtt::token &token)
{
    spdlog::warn("[tMqttClient] on_failure");
    spdlog::warn("[tMqttClient] token reason code {}", ReasonCodeToStdString(token.get_reason_code()));
    spdlog::warn("[tMqttClient] token message id {}", token.get_message_id());
    switch (token.get_type())
    {
    case mqtt::token::Type::CONNECT:
    {
        spdlog::warn("[tMqttClient] CONNECT FAILED, trying to reconnect");
        std::this_thread::sleep_for(std::chrono::milliseconds(5000));
        this->connect();
        break;
    }
    case mqtt::token::Type::DISCONNECT:
    {
        // Some logging
        break;
    }
    case mqtt::token::Type::PUBLISH:
    {
        spdlog::error("[tMqttClient] PUBLISH FAILED code={} mid={}",
                      ReasonCodeToStdString(token.get_reason_code()),
                      token.get_message_id());
        break;
    }
    case mqtt::token::Type::SUBSCRIBE:
    {
        // Some logging
        break;
    }
    case mqtt::token::Type::UNSUBSCRIBE:
    {
        // Some logging
        break;
    }
    default:
    {
        // Some logging
        break;
    }
    }
}
```

The last logging I see from the application is ```CONNECT FAILED, trying to reconnect```. So it seems that this exception when publishing is not using a callback.

**Comment by mbvlist** [comment]
I'm running into the same issue. I'm using async_client. I was surprised that the connect() invocation throws an exception at the first attempt, even though I specified automatic_reconnect in the passed options. If I start the broker, connect, and then publish a message, the same exception is thrown. 

I have a couple of questions:
* why is automatic_reconnect not handled after connect() is invoked?
* why is an exception thrown when publishing to a disconnected broker? 
* what happens with the data I attempted to publish using the async_client? Will it be published once the connection is re-established, or do I need to re-publish it?

I think the first 2 questions are actually bugs.

**Comment by alec-bell** [comment]
> The synchronous client just wraps an async one, so you can use callbacks with one, but that often just gets you into trouble with deadlocks. The more common usage is to detect a disconnect on a failed publish or consume operation and try to reconnect then. The `sync_consume.cpp` example shows this.

The suggestion here worked for me. After implementing it in my project with `mqtt::async_client` with automatic reconnect enabled, the solution looks like the following. I chose to not re-publish the message as it doesn't matter for my use-case, so please don't forget to add that logic in if your project needs it.
```
try {
    client->publish(topic, payload, size);
} catch (mqtt::exception& e) {
    while (!client->is_connected()) {
      std::this_thread::sleep_for(std::chrono::milliseconds(250));
    }
}
```

**Comment by fpagliughi** [comment]
There were a couple issues involving reconnect in the upstream C library that were fixed with Paho C v1.3.13. Anyone experiencing problems or crashes with reconnecting should update to that version.

The upcoming release of this C++ library will recommend that version and test with it.
